### PR TITLE
Handle maps with complex keys properly in fixValue

### LINKED ISSUE #2738: Handle maps with complex keys properly in fixValue
With the latest master when I create a map with complex keys I get an exception at the client side. 
Both of these queries fail:

```
select MAP(ARRAY [ARRAY[1,2,3,4]], ARRAY [ARRAY[2]]);
select MAP(ARRAY [MAP(ARRAY[1],ARRAY[2])], ARRAY [ARRAY[2]]);
```

```
Caused by: java.lang.IllegalArgumentException: Invalid [simple type, class com.facebook.presto.client.QueryResults] json bytes
        at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:174) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at io.airlift.http.client.FullJsonResponseHandler$JsonResponse.<init>(FullJsonResponseHandler.java:116) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        ... 11 common frames omitted
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Instantiation of [simple type, class com.facebook.presto.client.QueryResults] value failed: Cannot cast java.lang.String to java.util.List
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapException(StdValueInstantiator.java:405) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:234) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:167) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:398) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1064) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:264) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:124) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3066) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2262) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:171) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        ... 12 common frames omitted
Caused by: java.lang.ClassCastException: Cannot cast java.lang.String to java.util.List
        at java.lang.Class.cast(Class.java:3361) ~[na:1.8.0_25]
        at com.facebook.presto.client.QueryResults.fixValue(QueryResults.java:226) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.facebook.presto.client.QueryResults.fixValue(QueryResults.java:236) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.facebook.presto.client.QueryResults.fixData(QueryResults.java:208) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.facebook.presto.client.QueryResults.<init>(QueryResults.java:83) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at sun.reflect.GeneratedConstructorAccessor13.newInstance(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_25]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:408) ~[na:1.8.0_25]
        at com.fasterxml.jackson.databind.introspect.AnnotatedConstructor.call(AnnotatedConstructor.java:125) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:230) ~[presto-cli-0.101-SNAPSHOT-executable.jar:0.101-SNAPSHOT]
        ... 20 common frames omitted
```

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
Fixes #2738 

The client receives a json with ``"data":[[{"[1,2,3,4]":[2]}]]`` as jackson serializes the map using ``StdKeySerializer`` which writes the map key using ``key.toString`` resulting in ``[1,2,3,4]``. So in the ``fixValue`` method this PR handles strings like ``[1,2,3,4]`` and ``{1=2}`` (for maps with keys of type map).

### CODE REVIEW

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java [hunk 1]
```diff
@@ -223,7 +248,15 @@ private static Object fixValue(String type, Object value)
         TypeSignature signature = parseTypeSignature(type);
         if (signature.getBase().equals(ARRAY)) {
             List<Object> fixedValue = new ArrayList<>();
-            for (Object object : List.class.cast(value)) {
+            List<Object> list;
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: Introduce methods for this `fixList` and `fixMap`

**Comment by electrum** [comment]
Does the same issue apply to any non-string map key? JSON does not allow numbers as object keys.

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java [hunk 2]
```diff
@@ -212,6 +212,31 @@ public String toString()
...
     }
 
+    public static List<Object> toList(String arrayAsString)
+    {
+        int len = arrayAsString.length();
+        checkArgument(arrayAsString.charAt(0) == '[' && arrayAsString.charAt(len - 1) == ']', "Input string should have opening and closing brackets, e.g., [1,2,3]");
+
+        arrayAsString = arrayAsString.substring(1, len - 1);
+        List<Object> matches = new ArrayList<>();
+        Splitter.on(',').trimResults().split(arrayAsString).forEach(element -> matches.add(element));
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: How does this handle complex nested types that contain a comma?

**Comment by electrum** [comment]
Can you add tests for this?

**Comment by nezihyigitbasi** [comment]
@electrum The current parser doesn't handle nested complex structures properly -- it's very simple. The best way to parse these array/map strings is to use the ``org.json.JSONArray`` and ``org.json.JSONObject`` classes, I verified that they can handle arbitrarily nested structures. But this means we need to introduce a new dependency to the ``presto-client`` module. It becomes much easier to parse these array/map strings with these classes:

```java
JSONArray inputArray = new JSONArray("[[1], [2], [[3]]]");
JSONObject inputMap = new JSONObject("{ 1 = [[2]] , 3 = [[4]]}");
```

**Comment by haozhun** [comment]
@nezihyigitbasi In Presto, we use `org.fasterxml.jackson` for json serialization and deserializtion. I just checked and it can't parse an object like this: `{1=2}`. I discussed with @electrum. And we believe that the current behavior that relies on `toString` is broken. We are inclined to change the representation of `map` (possibly using a list of tuples), which will break backward compatibility. As a result, we are still considering possible alternatives.

**Comment by haozhun** [comment]
Superseded by #3755

**Pull request closed without merging**